src/Optimization/hiopAlgPrimalDecomp.cpp

```diff
@@ -1204,6 +1236,523 @@ void hiopAlgPrimalDecomposition::set_alpha_max(const double alp_max)
+            hiopVector* grad_aux = x0->alloc_clone();
+            grad_aux->setToZero(); 
+
+            for(int ri=0; ri<cont_idx.size(); ri++) {
+              int idx_temp = cont_idx[ri];
+              bret = master_prob_->eval_grad_rterm(idx_temp, nc_, x0_vec, *grad_aux);
+              if(!bret) {
+                //TODO
+              }
+              //grad_acc->axpy(1.0, *grad_aux);
+              grad_r->axpy(1.0, *grad_aux);
+            }
+
+            //rec_prob[my_rank_]->set_value(rec_val);
+            //rec_prob[my_rank_]->set_grad(grad_acc_vec);
+           
+            // send signal that the subproblem has been solved 
+            rec_prob[my_rank_]->post_send_end_signal(2, rank_master, comm_world_);
+            //rec_prob[my_rank_]->post_send(2, rank_master, comm_world_);
+            // log_->printf(hovIteration, "send recourse value flag for test %d \n", mpi_test_flag); 
+        
+            // post recv for new index
+            req_cont_idx->post_recv(1, rank_master, comm_world_);
+            // ierr = MPI_Irecv(&cont_idx[0], 1, MPI_INT, rank_master, 1, comm_world_, &request_[0]);
+            delete grad_aux;
+          }
+        }
+        //rval = rec_val;
+        //grad_r->copyFrom(grad_acc_vec);
+      }
+
+      if(my_rank_==0) {
+        assert(rval == 0);
+        for(int i=0; i<nc_; i++ ) {
+          assert(grad_r_vec[i] == 0.);
+        }
+        int mpi_test_flag = 0;
+        for(int r=1; r<comm_size_; r++) {
+          MPI_Wait(&(rec_prob[r]->request_), &status_);
+          MPI_Wait(&req_cont_idx->request_, &status_);
+        }
+      }
+
+      //std::cout<<"my rank "<<my_rank_<< " grad "<< grad_r_vec[0]<< " "<<grad_r_vec[1]<<std::endl;
+      MPI_Reduce(&rval, &rval_main, 1, MPI_DOUBLE, MPI_SUM, 0, comm_world_); // collect recourse function value 
+      MPI_Reduce(grad_r_vec, grad_r_main_vec, nc_, MPI_DOUBLE, MPI_SUM, 0, comm_world_); // collect recourse function gradient 
+      
+      if(my_rank_==0) {
+        
+        //std::cout<<"real rval %18.12e\n "<< rval_main<<std::endl;
+        rval = rval_main;
+        grad_r->copyFrom(grad_r_main_vec);
+
+        rval /= S_;
+        grad_r->scale(1.0/S_);
+
+        log_->printf(hovSummary, "real rval %18.12e\n", rval);
+        
+        MPI_Status mpi_status; 
+
+        for(int i=0; i<nc_; i++) {
+          hess_appx_vec[i] = 1.0;
+        }
+
+        if(nc_<n_) {
+          x0->copy_from_indexes(*x_, *xc_idx_);
+        } else {
+          assert(nc_==n_);
+          x0->copyFromStarting(0, *x_);
+        }
+
+        if(it==0) {
+          hess_appx_2->initialize(rval, *x0, *grad_r);
+          double alp_temp = hess_appx_2->get_alpha_f(*grad_r);
+          // double alp_temp = hess_appx_2->get_alpha_tr(); // alternative update rule for alpha
+          log_->printf(hovSummary, "alpd %18.12e\n", alp_temp);
+          
+          for(int i=0; i<nc_; i++) {
```

Comment:
`hess_appx->setToConstant(alp_temp)`

Reply:
This is also a problem in the run() function.